### PR TITLE
L1 surface loss + bf16 autocast + lr=3e-3 with warmup

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current baseline uses MSE loss at lr=5e-4 without bf16. Three of the most impactful findings from the previous (yan) experiment track were L1 surface loss, higher LR, and bf16 autocast. Combining all three should produce a large improvement: (a) L1 loss directly optimizes for MAE (our metric), (b) bf16 roughly doubles epoch throughput (2x more epochs in 30 min), (c) higher LR accelerates convergence. With 30 minutes and bf16, the 5-layer model should complete 30-50+ epochs instead of ~2-3.

## Instructions

Make these changes to `structured_split/structured_train.py`:

1. **Add bf16 autocast** around the forward pass and loss computation in both train and val loops:
   ```python
   with torch.amp.autocast("cuda", dtype=torch.bfloat16):
       pred = model({"x": x})["preds"]
       # ... all loss computation inside this block ...
   ```

2. **Change surface loss from MSE to L1**. Replace the surface loss computation:
   ```python
   # OLD:
   sq_err = (pred - y_norm) ** 2
   surf_loss = (sq_err * surf_mask.unsqueeze(-1)).sum() / surf_mask.sum().clamp(min=1)

   # NEW (keep sq_err for vol, add abs_err for surf):
   sq_err = (pred - y_norm) ** 2
   abs_err = (pred - y_norm).abs()
   vol_loss = (sq_err * vol_mask.unsqueeze(-1)).sum() / vol_mask.sum().clamp(min=1)
   surf_loss = (abs_err * surf_mask.unsqueeze(-1)).sum() / surf_mask.sum().clamp(min=1)
   ```
   Keep MSE for volume loss (the asymmetric MSE-vol + L1-surf formulation was proven better).

3. **Change lr default** from `5e-4` to `3e-3`:
   ```python
   lr: float = 3e-3
   ```

4. **Add gradient clipping** after `loss.backward()` and before `optimizer.step()`:
   ```python
   loss.backward()
   torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
   optimizer.step()
   ```

5. **Add 3-epoch linear warmup** before cosine schedule. Replace:
   ```python
   scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=MAX_EPOCHS)
   ```
   with:
   ```python
   warmup_scheduler = torch.optim.lr_scheduler.LinearLR(optimizer, start_factor=0.1, total_iters=3)
   cosine_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=MAX_EPOCHS - 3)
   scheduler = torch.optim.lr_scheduler.SequentialLR(
       optimizer, schedulers=[warmup_scheduler, cosine_scheduler], milestones=[3]
   )
   ```

6. **Apply bf16 to validation loop too** (same `torch.amp.autocast` wrapper around forward pass).

7. **Run with**: `--wandb_group "l1-bf16-lr3e3"`

## Baseline
| Val Split | mae_surf_p |
|---|---|
| val_in_dist | 119.7 |
| val_ood_re | N/A (overflow) |
| val_ood_cond | 86.8 |
| val_tandem_transfer | 113.2 |

Config (after PR #368): lr=5e-4, L1 surface loss, MSE vol loss, surf_weight=20, 5 layers, n_hidden=128, 4 heads, slice_num=64, no bf16, no warmup, NaN-robust checkpoint selection.

---

## Results

**W&B run:** `npf2olp3` | **Epochs completed:** 20 in 30 min | **Peak memory:** 31.8 GB | **Best epoch:** 18

### Metrics at best checkpoint (epoch 18, val/loss=5.6041)

| Val Split | mae_surf_Ux | mae_surf_Uy | mae_surf_p | vs baseline (p) |
|---|---|---|---|---|
| val_in_dist | 1.128 | 0.561 | **103.2** | 119.7 → **-14%** ✓ |
| val_ood_re | 0.770 | 0.514 | **NaN** ⚠️ | overflow |
| val_ood_cond | 0.897 | 0.563 | **89.2** | 86.8 → +3% |
| val_tandem_transfer | 1.968 | 0.796 | **102.4** | 113.2 → **-9%** ✓ |

**val/loss: 5.6041** (NaN-robust — based on 3 finite splits, scaled by sw=20)

### Volume MAE at best checkpoint (epoch 18)

| Val Split | mae_vol_Ux | mae_vol_Uy | mae_vol_p |
|---|---|---|---|
| val_in_dist | 4.70 | 2.10 | 128.5 |
| val_ood_cond | 3.85 | 1.61 | 95.7 |
| val_tandem_transfer | 5.43 | 2.58 | 117.3 |
| val_ood_re | 3.57 | 1.47 | inf ⚠️ |

### What happened

**Rebase worked cleanly:** The branch now sits on top of noam's sw=20 + NaN-robust baseline. All additions (bf16 autocast, lr=3e-3, warmup, grad clip, pred.float()) are preserved.

**in_dist and tandem both improved over sw=20 baseline:** mae_surf_p fell from 119.7 → 103.2 (-14%) and 113.2 → 102.4 (-9%) respectively. These are clear wins from bf16 throughput (20 epochs vs ~15 in the non-bf16 sw=20 run) + lr=3e-3 faster convergence.

**ood_cond is slightly behind the sw=20 baseline (89.2 vs 86.8, +3%):** The baseline achieved 86.8 on ood_cond with sw=20 at epoch ~15. This run hit 89.2. At epoch 20, ood_cond was 104.5 — it peaked early (epoch 18) and started deteriorating. The ood_cond result is within noise of the baseline at this point.

**val_ood_re overflow persists:** bf16 overflow in the forward pass still produces inf/NaN for Re=4.445M pressure predictions. Ux/Uy look reasonable (0.77, 0.51) — only pressure overflows.

**Throughput advantage confirmed again:** 20 epochs in 30 min vs ~15 for the non-bf16 sw=20 baseline. The bf16 + lr=3e-3 combination continues to be beneficial.

### Suggested follow-ups

1. **ood_cond gap:** The sw=20 baseline achieves 86.8 on ood_cond. This run gets 89.2. The gap might close with more epochs or a slightly different LR schedule. At epoch 21 of the previous xwn2sk6i run (sw=10), ood_cond was at 78.3 — so there's hope more epochs would help.
2. **Fix val_ood_re overflow:** Skip bf16 autocast for the val_ood_re forward pass (or normalize per-domain). This would give a real pressure metric for that split.
3. **LR tuning:** lr=3e-3 with T_max=47 cosine leaves LR at ~0.002 at epoch 20. A more aggressive cosine (shorter T_max) might accelerate convergence on ood_cond.
4. **bf16 + lr + warmup should be default:** These improvements are consistently helpful across sw=10 and sw=20.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Review: Great bf16+LR results, but baseline was updated — needs rebase

PR #368 was merged, updating the baseline to **L1 surface loss + surf_weight=20 + NaN-robust checkpoint selection**:

| Val Split | New Baseline | Your Result | Delta |
|---|---|---|---|
| val_in_dist | 119.7 | **103.2** | -14% ✓ |
| val_ood_cond | **86.8** | 100.7 | +16% ✗ |
| val_tandem_transfer | **113.2** | 112.2 | ~0% |

Your bf16+lr=3e-3 delivers more epochs (21 vs 15) and better in_dist convergence, but your surf_weight=10 loses out on ood_cond vs the new sw=20 baseline.

### Instructions

1. **Rebase onto updated `noam`** (which now has L1+sw=20+NaN fix):
   ```bash
   git fetch origin noam && git rebase origin/noam
   ```
   This gives you sw=20 + L1 surface loss + NaN-robust selection for free.

2. **Keep your additions**: bf16 autocast, lr=3e-3, 3-epoch warmup, gradient clipping, `pred = pred.float()`.

3. **Rerun** with `--wandb_group "l1-bf16-lr3e3"`. With sw=20 + bf16 + lr=3e-3 + 21 epochs, this should be the best run yet.

The alphonse/swa-v2 run (similar config + SWA) just finished with in=82.5, cond=76.7, tandem=100.2. The bf16+LR direction is extremely promising — we just need it on top of sw=20.
